dspy/utils/cache.py

```diff
@@ -0,0 +1,156 @@
+import os
+import ujson
+import pickle
+import litellm
+import pydantic
+import threading
+
+from pathlib import Path
+from hashlib import sha256
+from functools import wraps
+from typing import Any, Dict
+from cachetools import LRUCache
+from diskcache import FanoutCache
+from litellm.caching import Cache as litellm_cache
+
+
+DISK_CACHE_DIR = os.environ.get("DSPY_CACHEDIR") or os.path.join(Path.home(), ".dspy_cache")
+DISK_CACHE_LIMIT = int(os.environ.get("DSPY_CACHE_LIMIT", 3e10))  # 30 GB default
+MEM_CACHE_LIMIT = float(os.environ.get("DSPY_CACHE_LIMIT", float("inf")))  # unlimited by default
+
+# TODO: There's probably value in separating the limit for
+# the LM cache from the embeddings cache. Then we can lower the default 30GB limit.
+litellm.cache = litellm_cache(disk_cache_dir=DISK_CACHE_DIR, type="disk")
+
+if litellm.cache.cache.disk_cache.size_limit != DISK_CACHE_LIMIT:
+    litellm.cache.cache.disk_cache.reset('size_limit', DISK_CACHE_LIMIT)
+
+class Cache:
+    """
+    DSPy's caching interface. It provides 2 levels of caching (in the given order):
+    1. An in memory cache - cachetools' lrucache
+    2. A disk based cache - diskcache's fanoutcache
+    """
+    
+    def __init__(self, directory, disk_size_limit, mem_size_limit):
+        """
+        Args:
+            directory: The directory where the disk cache is stored.
+            disk_size_limit: The maximum size of the disk cache (in bytes).
+            mem_size_limit: The maximum size of the in-memory cache (in number of items).
+        """
+
+        self.memory_cache = LRUCache(maxsize=mem_size_limit)
+        self.fanout_cache = FanoutCache(shards=16, timeout=2, directory=directory, size_limit=disk_size_limit)
```

Comment:
@hmoazam As I understand it, part of the motivation for this PR is to enable users to dump the current state of the cache for a particular DSPy session.

Is the `FanoutCache` introduced for that purpose? If so, since `FanoutCache` shares the same `directory` as the `litellm.cache`, what extra utility is it providing?

Reply:
Separately, I think we should define the behavior for dumping / loading a cache. 

For example, consider the case where I've already run "Program 1" as a Python script / notebook. Now, I'm running "Program 2" in a separate notebook / Python script invocation. I'm using the default DSPy cache directory for both program executions.

If I dump the cache after I run Program 2, do I also want to include cache contents from Program 1? Or, do I only want to include cache contents from Program 1, since that's my "current session"? Should users have a choice?

Once we align on the desired semantics, we can figure out how to make adjustments to the implementation.

cc @okhat

Reply:
@dbczumar For disk caching, we will migrate from LiteLLM's caching to FanoutCache. The former has proved limiting when users need many read/write processes and threads in parallel.

For a short period of time, we might keep the LiteLLM cache _as well_. This means that highly active users that upgrade versions regularly will have an implicit seamless migration of their caches, because LiteLLM's requests will remain cached and will implicitly transfer to the FanoutCache. (This behavior is not required, and in principle there are other ways to more intentionally allow migration, but explicit migration is almost never worthwhile for caches.)

Cache dumping is fairly simple. It saves the current in-memory Python session, unless the user triggers a reset for the in-memory cache. (The current in-memory session draws on saved caches on disk, as usual.)